cirq-core/cirq/transformers/transformer_primitives.py

```diff
@@ -328,22 +367,35 @@ def unroll_circuit_op_greedy_earliest(
 
     Args:
         circuit: Input circuit to apply the transformations on. The input circuit is not mutated.
+        deep: If True, `unroll_circuit_op_greedy_earliest` is recursively called on all circuit
+            operations matching `tags_to_check`.
         tags_to_check: If specified, only circuit operations tagged with one of the `tags_to_check`
             are unrolled.
 
     Returns:
         Copy of input circuit with (Tagged) CircuitOperation's expanded using EARLIEST strategy.
     """
     batch_removals = [*circuit.findall_operations(lambda op: _check_circuit_op(op, tags_to_check))]
-    batch_inserts = [(i, protocols.decompose_once(op)) for i, op in batch_removals]
+    batch_inserts = []
+    for i, op in batch_removals:
+        sub_circuit = cast(circuits.CircuitOperation, op.untagged).mapped_circuit()
+        sub_circuit = (
+            unroll_circuit_op_greedy_earliest(sub_circuit, deep=deep, tags_to_check=tags_to_check)
```

Comment:
Can we modify the CircuitOperations without explicitly unrolling them ?

Reply:
The method, `unroll_circuit_op_greedy_earliest`, is supposed to unroll all circuit operations that match the tags. If `deep = True` is specified, the unrolling is supposed to happen recursively so that all nested circuit operations satisfying the criteria are also explicitly unrolled. 

Hence, unrolling circuit operations explicitly is the intended behaviour here.

Reply:
Sort of. That seems pretty slow though in some cases, I could imagine instances where the user might just want to update optimize the circuitoperation without the unrolling happening. Would having this disallow them to do that ?

Reply:
The `tags_to_ignore` flag is used to enable that. For example:
- `map_operations(circuit, map_func, deep=True)` can be used to recursively optimise the circuit  + circuit operations inside circuit. The newly added mapped operations will be wrapped in a tagged circuit op by default. 
- `unroll_circuit_op(circuit, deep=True)` will, by default, recursively unroll only tagged circuit operations introduced in the above step.  

Thus, combination of above two steps will enable users to optimize circuit operations without unrolling them.